### PR TITLE
1 emloyee screen backend

### PR DESCRIPTION
**Pull request by imujo-fer**

### CODE REVIEW

File: src/main/java/com/progi/company/Company.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        this.eurCostPerKm = eurCostPerKm;
+        this.locationCoordLat = locationCoordLat;
+        this.locationCoordLon = locationCoordLon;
+        this.address = address;
+        this.city = city;
+        this.country = country;
+        this.iban = iban;
+    }
+
+    public Integer getId() {
```
Comment (imujo-fer) [inline src/main/java/com/progi/company/Company.java:69]: Provjeri rade li @Getter, setter, no args i all args constructori i ako rade makni runco napisane stvari na svim entitetima i dtoovima

File: src/main/java/com/progi/expensereport/ExpenseReport.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+public class ExpenseReport {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Integer id;
+
+    @OneToOne
+    @JoinColumn(name = "trip_id", referencedColumnName = "id")
+    private Trip trip;
+
+    private Double eurTotalCost;//nullable
```
Comment (imujo-fer) [inline src/main/java/com/progi/expensereport/ExpenseReport.java]: Maknuti komentar

File: src/main/java/com/progi/expensereport/ExpenseReportController.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+@RequestMapping("/api/expense-reports")
+public class ExpenseReportController {
+
+    @Autowired
+    private ExpenseReportService expenseReportService;
+
+
+    @GetMapping("/{id}/items")
+    public ResponseEntity<List<ExpenseReportItemWithSubcategoryDTO>> getExpenseReportItems(@PathVariable Integer id) {
+        ExpenseReport expenseReport = expenseReportService.getExpenseReportById(id);
```
Comment (imujo-fer) [inline src/main/java/com/progi/expensereport/ExpenseReportController.java]: Pomaknuti ovu logiku u servis

File: src/main/java/com/progi/trip/Trip.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    @OneToOne(mappedBy = "trip")
+    @JsonManagedReference
+    private ExpenseReport expenseReport;
+
+    @OneToMany(mappedBy = "trip")
+    @JsonManagedReference
+    private List<TripStatus> tripStatuses;
+
+    @Column(nullable = false)
+    private Timestamp createdAt; // This will be set in @PrePersist
```
Comment (imujo-fer) [inline src/main/java/com/progi/trip/Trip.java]: maknuti komentar

File: src/main/java/com/progi/trip/Trip.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+@Getter
+@Setter
+
+public class Trip {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Integer id;
+
+    @Column(unique = true, nullable = true)
+    private String requestNumber; // auto generated
```
Comment (imujo-fer) [inline src/main/java/com/progi/trip/Trip.java]: Tu treba promijenit u nullable false i popravit da se auto generira

File: src/main/java/com/progi/trip/TripController.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        tripDTO.setUser(new UserDTO(user));
+
+        if (trip.getUser() != null && trip.getUser().equals(user)) {
+            return ResponseEntity.ok(tripDTO);
+        }
+
+        if (user != null && user.getDepartment() != null) {
+            List<User> departmentHeads = departmentService.getDepartmentHeadsByDepartmentId(user.getDepartment().getId());
+
+            // Provjeravamo da li je korisnik jedan od department heads
```
Comment (imujo-fer) [inline src/main/java/com/progi/trip/TripController.java]: bool isUserDepartmentHead = ....

if (isUserDepartmentHead)

File: src/main/java/com/progi/trip/TripController.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    @GetMapping("/employee/{employeeID}")
+    public ResponseEntity<Page<TripResponseDTO>> getEmployeeTripsByStatus(
+            @PathVariable int employeeID,
+            @RequestParam(required = false) Status status,
+            @RequestParam(defaultValue = "0") int page,
+            @RequestParam(defaultValue = "10") int size) {
+
+        PageRequest pageRequest = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "createdAt"));
```
Comment (imujo-fer) [inline src/main/java/com/progi/trip/TripController.java]: Prebaciti u servis

File: src/main/java/com/progi/tripstatus/TripStatusRepository.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface TripStatusRepository extends JpaRepository<TripStatus, Integer> {
+    @Query("SELECT ts FROM TripStatus ts WHERE ts.trip.id = :tripId ORDER BY ts.createdAt DESC LIMIT 1")
+    Optional<TripStatus> findTopByTripIdOrderByCreatedAtDesc(@Param("tripId") Integer tripId);
```
Comment (imujo-fer) [inline src/main/java/com/progi/tripstatus/TripStatusRepository.java]: Ovo maknut

File: src/main/java/com/progi/trip/TripService.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        return tripRepository.findAll();
+    }
+
+    public Trip getTripById(Integer id) {
+        return tripRepository.findById(id).orElseThrow(() -> new NoSuchElementException("Trip not found with id " + id));
+
+
+    }
+
+    public Trip createTrip(TripDTO tripDetails) {
```
Comment (imujo-fer) [inline src/main/java/com/progi/trip/TripService.java:52]: dodat logiku da se nakon kreiranja tripa kreira i prvi status

File: src/main/java/com/progi/trip/TripController.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        return ResponseEntity.status(403).build();
+    }
+
+    @PostMapping
+    public ResponseEntity<Trip> createTrip(@RequestBody TripDTO trip) {
+        Trip createdTrip = tripService.createTrip(trip);
+        return ResponseEntity.ok(createdTrip);
+    }
+
+    @PatchMapping("/{id}")
```
Comment (imujo-fer) [inline src/main/java/com/progi/trip/TripController.java]: Preimenovat sve PathMapping u PutMapping

File: src/main/java/com/progi/trip/TripController.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+        TripWithCountryDTO tripDTO = new TripWithCountryDTO(trip);
+        tripDTO.setUser(new UserDTO(user));
+
+        if (trip.getUser() != null && trip.getUser().equals(user)) {
+            return ResponseEntity.ok(tripDTO);
+        }
+
+        if (user != null && user.getDepartment() != null) {
+            List<User> departmentHeads = departmentService.getDepartmentHeadsByDepartmentId(user.getDepartment().getId());
```
Comment (imujo-fer) [inline src/main/java/com/progi/trip/TripController.java]: Ovu logiku stavit u entity. Takoder mozemo slicno napravit za sve role isUserDepHead...

File: src/main/java/com/progi/receipt/Receipt.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Integer id;
+
+    @Column(nullable = false)
+    @NotNull
+    private String path;
+
+    @OneToOne(mappedBy = "receipt")
+    private ExpenseReportItem expenseReportItem;
```
Comment (imujo-fer) [inline src/main/java/com/progi/receipt/Receipt.java:29]: Dodat json ignore

File: src/main/java/com/progi/expensereportitem/ExpenseReportItemWithSubcategoryDTO.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    private Currency currency;
+    @NotNull
+    private Double currencyValue;
+    @NotNull
+    private Double eurValue;
+
+    public ExpenseReportItemWithSubcategoryDTO(ExpenseReportItem expenseReportItem) {
+        this.expenseReportId = expenseReportItem.getExpenseReport().getId();
+        this.receiptId = expenseReportItem.getReceipt().getId();
+        this.expenseSubcategory = expenseReportItem.getExpenseSubcategory();
```
Comment (imujo-fer) [inline src/main/java/com/progi/expensereportitem/ExpenseReportItemWithSubcategoryDTO.java:28]: Ovi subcategories ne vracaju i category unutar sebe. Tu ce mozda trebat radit neki dto

File: src/main/java/com/progi/expensereportitem/ExpenseReportItemService.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+    @Autowired
+    private ReceiptService receiptService;
+    @Autowired
+    private ExpenseSubcategoryService expenseSubcategoryService;
+
+    public ExpenseReportItem getExpenseReportItemById(Integer id) {
+        return expenseReportItemRepository.findById(id).orElseThrow(() -> new NoSuchElementException("Expense report item not found with id " + id));
+    }
+
+    public void deleteExpenseReportItem(Integer id) {
```
Comment (imujo-fer) [inline src/main/java/com/progi/expensereportitem/ExpenseReportItemService.java:32]: Treba deleteat receipt, ne treba deleteat expense report subcategory

File: src/main/java/com/progi/company/Company.java (hunk 1 above)
Comment (imujo-fer) [reply src/main/java/com/progi/company/Company.java:69]: 3130c74b521b22446d7d11e47ac0bea41718d8fd

Comment (adrianofaletar) [reply src/main/java/com/progi/company/Company.java:69]: f1e55e961d07fde3678044ad817ab8fb12a51f31

File: src/main/java/com/progi/expensereport/ExpenseReport.java (hunk 2 above)
Comment (adrianofaletar) [reply src/main/java/com/progi/expensereport/ExpenseReport.java]: 9b3a99cfdbb1b239c035c17927e12297b1c7ceb3

File: src/main/java/com/progi/expensereport/ExpenseReportController.java (hunk 3 above)
Comment (adrianofaletar) [reply src/main/java/com/progi/expensereport/ExpenseReportController.java]: af5f76dcfbc42438f5a8cf4381a2cd81466019a6

File: src/main/java/com/progi/trip/Trip.java (hunk 4 above)
Comment (adrianofaletar) [reply src/main/java/com/progi/trip/Trip.java]: komentar je maknut u nekom drugom commitu

File: src/main/java/com/progi/trip/Trip.java (hunk 5 above)
Comment (adrianofaletar) [reply src/main/java/com/progi/trip/Trip.java]: napravljeno u drugom commitu

File: src/main/java/com/progi/trip/TripController.java (hunk 10 above)
Comment (adrianofaletar) [reply src/main/java/com/progi/trip/TripController.java]: 71c834c0ed5a0577189e096a7b33cdb9e2ea5f4e

File: src/main/java/com/progi/trip/TripController.java (hunk 11 above)
Comment (adrianofaletar) [reply src/main/java/com/progi/trip/TripController.java]: 4c05b2d7efc47f992d81257e28bcb8905f463870

File: src/main/java/com/progi/receipt/Receipt.java (hunk 12 above)
Comment (adrianofaletar) [reply src/main/java/com/progi/receipt/Receipt.java:29]: napravljeno u drugom commitu

File: src/main/java/com/progi/expensereportitem/ExpenseReportItemService.java (hunk 14 above)
Comment (adrianofaletar) [reply src/main/java/com/progi/expensereportitem/ExpenseReportItemService.java:32]: a204136791a4aaa11f166db9570e4eb1894b6593

Comment (imujo) [reply src/main/java/com/progi/expensereportitem/ExpenseReportItemService.java:32]: Mozemo vjerojatno ovo koristit @OnDelete(action = OnDeleteAction.CASCADE)

File: src/main/java/com/progi/trip/TripController.java (hunk 7 above)
Comment (adrianofaletar) [reply src/main/java/com/progi/trip/TripController.java]: 6c904fd2e5e7eeb9b2916a37a71a8a73b4c8ee52

File: src/main/java/com/progi/tripstatus/TripStatusRepository.java (hunk 8 above)
Comment (adrianofaletar) [reply src/main/java/com/progi/tripstatus/TripStatusRepository.java]: 32c8742cedc8457f0348c74f869b49252634d86c

File: src/main/java/com/progi/trip/TripService.java (hunk 9 above)
Comment (adrianofaletar) [reply src/main/java/com/progi/trip/TripService.java:52]: 64458fff114a5cf86ffd32ee384d52e5fcf105f3

**Pull request merged**